src/ai-services/ant-eater.service.ts

```diff
@@ -0,0 +1,127 @@
+
+    constructor(
+        private readonly configService: ConfigService,
+        private readonly messageService: MessageService,
+    ) {
+        super();
+    }
+
+    private async updateChatHistoryCache(chatId: number) {
+        const { HumanMessage, AIMessage } = await dynamicImport('langchain/schema');
+        const pastMessages = [];
+            const messages = await this.messageService.findAll(chatId);
+            messages.forEach((m) => {
+                if (m.role === 'user') {
+                    pastMessages.push(new HumanMessage(m.content));
+                } else if (m.role === 'assistant') {
+                    pastMessages.push(new AIMessage(m.content));
+                }
+            });
+        Console.log(typeof pastMessages)
+        this.chatHistoryCache.set(chatId, pastMessages);
+        Console.log("Finished updated")
+        for (let [key, value] of this.chatHistoryCache) {
+        }
+    }
+
+    public async generateResponse(
+        input: string,
+        chatId?: number,
+    ): Promise<string> {
+        const { OpenAI } = await dynamicImport('langchain/llms/openai');
+        const { ConversationSummaryMemory, BufferMemory, ConversationSummaryBufferMemory, CombinedMemory, ChatMessageHistory } =
+            await dynamicImport('langchain/memory');
+        const { LLMChain } = await dynamicImport('langchain/chains');
+        const { PromptTemplate } = await dynamicImport('langchain/prompts');
+
+        // Retrieve or update chat history from cache
+        if (!this.chatHistoryCache.has(chatId)) {
+            await this.updateChatHistoryCache(chatId);
+        }
+        const pastMessages: Message[] = this.chatHistoryCache.get(chatId) || [];
+        // Initialize OpenAI model
+        const model = new OpenAI({
+            openAIApiKey: this.configService.get<string>('OPENAI_API_KEY'),
+            modelName: 'gpt-3.5-turbo-16k',
+            temperature: 0,
+            streaming: false,
+        });
+
+        const bufferMemory = new BufferMemory({
+            memoryKey: "chat_history",
+            inputKey: "input",
+            chatHistory: new ChatMessageHistory(this.chatHistoryCache.get(chatId)),
+        });
+
+        const summaryMemory = new ConversationSummaryMemory({
+            llm: model,
+            maxTokenLimit: 50,
+            memoryKey: "conversationSummary",
+            inputKey: "input",
+
+        });
+
+        const memory = new CombinedMemory({
+            memories: [bufferMemory, summaryMemory],
+        });
+        Console.log("Created memory objects")
+
+        // Prompt Template
+        const prompt = PromptTemplate.fromTemplate(`The following is a friendly conversation between a human and an AI. 
+The AI is talkative and provides lots of specific details from its context. 
+If the AI does not know the answer to a question, it truthfully says it does not know.
+
+Summary of conversation:
+{conversationSummary}
```

Comment:
The summary of the conversation is never being added here. Flip the `verbose: false`, parameter of `LLMChain` to `true` to see what goes in here.